src/java/com/puppetlabs/jruby_utils/pool/JRubyPool.java

```diff
@@ -202,31 +203,50 @@ public E borrowItemWithTimeout(long timeout, TimeUnit unit) throws
         return item;
     }
 
+    /**
+     * Release an item and return it to the pool. Does nothing if the item
+     * being released is the pill.
+     * Throws an `IllegalArgumentException` if the item is not currently
+     * registered by the pool and the item is not the pill, if one has been
+     * inserted
+     */
     @Override
     public void releaseItem(E e) {
         final ReentrantLock lock = this.queueLock;
         lock.lock();
         try {
-            // Do nothing if the item being returned is the poisonPill
-            if (e != this.poisonPill){
+            if (!isRegistered(e) && e != this.pill){
```

Comment:
Could just do the `!isRegistered(e)` check and throw an exception inside of the `e != this.pill` check around the `addFirst(e)` call below.  That way, you'd only need to do the `e != this.pill` once.